language/move-vm/runtime/src/loader.rs

```diff
@@ -990,35 +1020,63 @@ impl Loader {
         bundle_verified: &BTreeMap<ModuleId, CompiledModule>,
         bundle_unverified: &BTreeSet<ModuleId>,
         data_store: &impl DataStore,
-    ) -> VMResult<Arc<Module>> {
-        // if the module is already in the code cache, load the cached version
-        if let Some(cached) = self.module_cache.read().module_at(id) {
-            self.module_cache_hits.write().insert(id.clone());
-            return Ok(cached);
-        }
+    ) -> VMResult<(Arc<CompiledModule>, Arc<Module>)> {
+        let compiled = {
+            let locked_cache = self.module_cache.read();
+            match (
+                locked_cache.compiled_module_at(id),
+                locked_cache.loaded_module_at(id),
+            ) {
+                (None, Some(_loaded)) => {
+                    unreachable!("Loaded module without verified compiled module");
```

Comment:
this is ok but something we may want to keep in mind. A binary module may exist without its compiled module being around (though a future to be defined). But given we have split the `CompiledModule` and `Module` in the API (both returned) we may want to consider that option. Otherwise it could make sense to return a `CompiledModule` from a `BinaryModule`/`Module`